bindings/rust/wasmedge-sys/src/vm.rs

```diff
@@ -0,0 +1,174 @@
+use super::wasmedge;
+use crate::{
+    raw_result::{decode_result, ErrReport},
+    string::StringRef,
+    value::Value,
+    wasi,
+};
+use std::os::raw::c_char;
+
+// If there is a third-party sdk based on wasmedge-sys
+// the private encapsulation  here can force the third-party sdk to use the api
+#[derive(Debug)]
+pub struct Vm {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_VMContext,
+    _private: (),
+}
+
+impl Vm {
+    pub fn create(config: &crate::config::Config) -> Result<Self, ErrReport> {
+        let ctx = unsafe {
+            wasmedge::WasmEdge_VMCreate(config.ctx, std::ptr::null_mut() /* store */)
+        };
+        if ctx.is_null() {
+            panic!("WasmEdge VM create failed");
+        }
+        Ok(Self { ctx, _private: () })
+    }
+
+    pub fn load_wasm_from_ast_module(
+        self,
+        module: &crate::module::Module,
+    ) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMLoadWasmFromASTModule(
+                self.ctx, module.ctx,
+            ))?;
+        }
+        Ok(self)
+    }
+
+    pub fn validate(self) -> Result<Self, ErrReport> {
+        unsafe {
+            decode_result(wasmedge::WasmEdge_VMValidate(self.ctx))?;
+        }
+        Ok(self)
+    }
+
+    pub fn instantiate(self) -> Result<Self, ErrReport> {
```

Comment:
is there really any case that you'd do these individually and not always apply all of them?

Reply:
This is only the sys library, the final application in the upper sdk, split out to facilitate the handling of errors at each step